lib/Conversion/TorchToTosa/TorchBackendLegalToTosaCustom.cpp

```diff
@@ -0,0 +1,201 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#include "torch-mlir/Conversion/TorchToTosa/TorchToTosa.h"
+#include "torch-mlir/Conversion/TorchToTosa/TosaLegalizeCommon.h"
+#include "torch-mlir/Conversion/TorchToTosa/TosaLegalizeUtils.h"
+
+#include "../PassDetail.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Dialect/Traits.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+#include "torch-mlir/Dialect/Torch/Utils/Utils.h"
+#include "torch-mlir/Dialect/TorchConversion/IR/TorchConversionDialect.h"
+#include "torch-mlir/Dialect/TorchConversion/Transforms/BackendTypeConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+
+namespace {
+
+class ConvertBackendLegalAtenOp : public ConversionPattern {
+public:
+  ArrayRef<std::string> customOps;
+
+  ConvertBackendLegalAtenOp(TypeConverter &typeConverter, MLIRContext *context,
+                            ArrayRef<std::string> customOps)
+      : ConversionPattern(typeConverter, MatchAnyOpTypeTag{}, 0, context) {
+    this->customOps = customOps;
+  }
+
+  Value matchCustomOpOperandValue(Value inputOperand, PatternRewriter &rewriter,
```

Comment:
I think it would be clearer to call this `convertOperandToTensor` and have the op return `nullptr` if the final value is not a tensor type. At the moment if an argument is an unsupported type, for example a string, there is no checks that would result in an error. All paths left as TODOs should result in errors with clear error messages

Reply:
So this is a little tricky, since any operand that is not one of the special constant cases (like `torch.constant.bool` or `torch.constant.int`) that should be matched and absorbed by the rewriter needs to pass through. The reason is, if an operand is a result of another operation then it should be handled by the Torch to TOSA conversion pass.

Now to handle these special operations, I think we need to have all of them specifically listed here in the switch-case. I updated the switch-case to include all of the operation types that _should be_ handled here. This way, any unhandled constant operand will return `nullptr` and the rewriter will `notifyMatchFailure`; otherwise it will pass through and the expectation is that Torch to TOSA conversion pass will need to convert it.

Does this look reasonable?

Reply:
> So this is a little tricky, since any operand that is not one of the special constant cases (like `torch.constant.bool` or `torch.constant.int`) that should be matched and absorbed by the rewriter needs to pass through. The reason is, if an operand is a result of another operation then it should be handled by the Torch to TOSA conversion pass.

I'm not sure I understand. Isn't the result type of every value returned by this function supposed to be a `RankedTensorType`?